File: solutions/2026-01-19/python/2026-01-19-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.py
```python
# Problem: Maximum Side Length of a Square with Sum Less than or Equal to Threshold
# Problem Summary: Find the largest square subgrid whose elements sum up to a value
#                  less than or equal to a given threshold.
# Link: https://leetcode.com/problems/maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/
#
# Approach:
# The problem asks for the maximum side length. This suggests that we can use binary search
# on the possible side lengths. For a given side length 'k', we need to efficiently check
# if there exists any square subgrid of size k x k whose sum is less than or equal to
# the threshold.
#
# To efficiently calculate the sum of any square subgrid, we can use a 2D prefix sum array
# (also known as an integral image).
# Let `prefix_sum[i][j]` store the sum of all elements in the rectangle from (0, 0) to (i-1, j-1).
# The sum of a square subgrid with top-left corner (r1, c1) and bottom-right corner (r2, c2)
# can be calculated as:
# sum(r1, c1, r2, c2) = prefix_sum[r2+1][c2+1] - prefix_sum[r1][c2+1] - prefix_sum[r2+1][c1] + prefix_sum[r1][c1]
#
# For a square of side length `k` and top-left corner (r, c), the bottom-right corner will be
# (r + k - 1, c + k - 1).
#
# The binary search will operate on the range of possible side lengths, from 0 to min(m, n).
#
# The `check(k)` function will iterate through all possible top-left corners (r, c)
# for a square of side length `k`. For each potential square, it calculates its sum using
# the prefix sum array and checks if the sum is <= threshold. If such a square is found,
# `check(k)` returns True.
#
# Binary Search Logic:
# low = 0, high = min(m, n) + 1 (exclusive upper bound for convenience)
# while low < high:
#   mid = low + (high - low) // 2
#   if mid == 0:  # A square of side 0 always exists and has sum 0.
#       low = mid + 1
#       continue
#   if check(mid):
#       # If a square of side 'mid' exists, we try for a larger side.
#       # The answer could be 'mid' or larger.
#       low = mid + 1
#   else:
#       # If no square of side 'mid' exists, we need to try smaller sides.
#       high = mid
#
# The final answer will be `low - 1`.
#
# Time Complexity:
# - Building the 2D prefix sum array: O(m * n)
# - Binary search: The range of side lengths is from 0 to min(m, n). Let S = min(m, n).
#   The binary search performs log(S) iterations.
# - Inside the binary search, the `check(k)` function iterates through possible top-left
#   corners of squares of side `k`. There are approximately (m - k + 1) * (n - k + 1) such corners.
#   For each corner, calculating the sum is O(1) using the prefix sum.
#   So, `check(k)` is roughly O(m * n).
# - Total time complexity: O(m * n * log(min(m, n)))
#
# Space Complexity:
# - Storing the 2D prefix sum array: O(m * n)
#
class Solution:
    def maximalSquare(self, mat: list[list[int]], threshold: int) -> int:
        m = len(mat)
        n = len(mat[0])

        # 1. Build the 2D prefix sum array
        # prefix_sum[i][j] will store the sum of all elements in the rectangle
        # from (0, 0) to (i-1, j-1) in the original matrix.
        # We add an extra row and column of zeros to simplify calculations.
        prefix_sum = [[0] * (n + 1) for _ in range(m + 1)]
        for r in range(m):
            for c in range(n):
                prefix_sum[r + 1][c + 1] = mat[r][c] + prefix_sum[r][c + 1] + prefix_sum[r + 1][c] - prefix_sum[r][c]

        # Helper function to get the sum of a square subgrid
        # with top-left corner (r1, c1) and bottom-right corner (r2, c2).
        # Note: In the prefix_sum array, indices are 1-based for convenience.
        # So, matrix coordinates (r, c) correspond to prefix_sum indices (r+1, c+1).
        def get_square_sum(r1, c1, r2, c2):
            # Convert matrix 0-based indices to prefix_sum 1-based indices
            # The sum of a rectangle from (r1, c1) to (r2, c2) (inclusive)
            # is prefix_sum[r2+1][c2+1] - prefix_sum[r1][c2+1] - prefix_sum[r2+1][c1] + prefix_sum[r1][c1]
            return prefix_sum[r2 + 1][c2 + 1] - prefix_sum[r1][c2 + 1] - prefix_sum[r2 + 1][c1] + prefix_sum[r1][c1]

        # Helper function to check if there exists any square of side length 'k'
        # whose sum is less than or equal to the threshold.
        def check(k):
            if k == 0: # A square of side 0 always exists and its sum is 0.
                return True
            # Iterate through all possible top-left corners of a square of side 'k'.
            # The top-left corner (r, c) means the square spans from (r, c) to (r+k-1, c+k-1).
            for r in range(m - k + 1):
                for c in range(n - k + 1):
                    # Calculate the sum of the current k x k square
                    square_sum = get_square_sum(r, c, r + k - 1, c + k - 1)
                    if square_sum <= threshold:
                        return True # Found a square of side k with sum <= threshold
            return False # No such square of side k was found

        # 2. Binary search for the maximum side length
        # The side length can range from 0 to min(m, n).
        # 'low' will be the smallest possible side length that might work or larger.
        # 'high' will be the smallest side length that is definitely too large or the boundary.
        low = 0
        high = min(m, n) + 1 # Upper bound is exclusive, so we add 1

        ans = 0 # Initialize answer to 0

        while low < high:
            mid = low + (high - low) // 2 # Calculate middle side length

            if check(mid):
                # If a square of side 'mid' exists and satisfies the condition,
                # it means 'mid' is a possible answer, and we try to find a
                # larger side. So, we update 'low' to 'mid + 1'.
                # We also update 'ans' because 'mid' is a valid side length.
                ans = mid
                low = mid + 1
            else:
                # If no square of side 'mid' satisfies the condition,
                # it means 'mid' is too large, and we need to search in the
                # smaller half. So, we update 'high' to 'mid'.
                high = mid

        # The binary search ensures that 'ans' stores the largest 'mid' for which check(mid) was True.
        # If no square with sum <= threshold exists (even of size 1), 'ans' will remain 0.
        return ans

```

File: solutions/2026-01-19/java/2026-01-19-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.java
```java
/**
 * LeetCode Problem: Maximum Side Length of a Square with Sum Less than or Equal to Threshold
 * Link: https://leetcode.com/problems/maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/
 *
 * Problem Summary:
 * Given a 2D matrix and a threshold, find the largest square submatrix whose sum of elements
 * is less than or equal to the given threshold.
 *
 * Approach:
 * The problem asks for the maximum side length. This suggests that if a square of side 'k' satisfies
 * the condition, then any square of side 'j' < 'k' also satisfies it (since it's a sub-square, and elements are non-negative).
 * This monotonicity allows us to use binary search on the possible side lengths.
 *
 * The possible side lengths range from 0 to min(m, n).
 * For a given side length `k`, we need an efficient way to calculate the sum of all possible squares of size `k x k`.
 * This can be done using a 2D prefix sum array (also known as an integral image).
 *
 * The 2D prefix sum `prefixSum[i][j]` will store the sum of all elements in the rectangle from (0,0) to (i-1, j-1).
 * The sum of a rectangle with top-left corner (r1, c1) and bottom-right corner (r2, c2) can be calculated as:
 * sum(r1, c1, r2, c2) = prefixSum[r2+1][c2+1] - prefixSum[r1][c2+1] - prefixSum[r2+1][c1] + prefixSum[r1][c1].
 *
 * For a square of side `k` with its bottom-right corner at `(r, c)`, the top-left corner will be at `(r-k+1, c-k+1)`.
 * So, the sum of such a square is `prefixSum[r+1][c+1] - prefixSum[r-k+1][c+1] - prefixSum[r+1][c-k+1] + prefixSum[r-k+1][c-k+1]`.
 *
 * The `check(sideLength)` function will iterate through all possible bottom-right corners of squares with the given `sideLength`
 * and use the prefix sum array to calculate the sum. If any square's sum is less than or equal to the `threshold`, it returns `true`.
 *
 * Binary Search:
 * - `low = 0`, `high = min(m, n)` (maximum possible side length).
 * - `ans = 0` (to store the maximum valid side length found so far).
 * - While `low <= high`:
 *     - `mid = low + (high - low) / 2`
 *     - If `check(mid)` is true:
 *         - A square of side `mid` is possible. Try for a larger side.
 *         - `ans = mid`
 *         - `low = mid + 1`
 *     - Else (`check(mid)` is false):
 *         - A square of side `mid` is too large. Try for a smaller side.
 *         - `high = mid - 1`
 * - Return `ans`.
 *
 * Time Complexity:
 * - Building the prefix sum array: O(m*n)
 * - Binary search performs log(min(m, n)) iterations.
 * - Inside each iteration, the `check(k)` function iterates through all possible bottom-right corners of squares of size `k x k`.
 *   This is approximately O(m*n).
 * - Total Time Complexity: O(m*n * log(min(m, n)))
 *
 * Space Complexity:
 * - For the 2D prefix sum array: O(m*n)
 * - Total Space Complexity: O(m*n)
 */
class Solution {
    public int maxSideLength(int[][] mat, int threshold) {
        int m = mat.length;
        int n = mat[0].length;

        // Create a 2D prefix sum array.
        // prefixSum[i][j] stores the sum of elements in the rectangle from (0,0) to (i-1, j-1).
        // The size is (m+1) x (n+1) to handle 1-based indexing for easier calculation.
        int[][] prefixSum = new int[m + 1][n + 1];

        // Populate the prefix sum array.
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                prefixSum[i + 1][j + 1] = mat[i][j] + prefixSum[i][j + 1] + prefixSum[i + 1][j] - prefixSum[i][j];
            }
        }

        // Binary search for the maximum side length.
        int low = 0;
        int high = Math.min(m, n); // The maximum possible side length is limited by the smaller dimension.
        int ans = 0; // Stores the maximum side length found that satisfies the condition.

        // While the search space is valid.
        while (low <= high) {
            int mid = low + (high - low) / 2; // Calculate the middle side length to check.

            // If mid is 0, it's always a valid square (empty or single element with sum <= threshold if allowed).
            // We want to find the largest non-zero side if possible.
            if (mid == 0) {
                ans = Math.max(ans, mid);
                low = mid + 1;
                continue;
            }

            // Check if a square of side 'mid' exists with sum <= threshold.
            if (check(prefixSum, m, n, mid, threshold)) {
                // If a square of side 'mid' is valid, it means we might be able to find a larger square.
                ans = mid; // Update the answer to the current valid side length.
                low = mid + 1; // Try searching in the upper half for a larger side length.
            } else {
                // If a square of side 'mid' is not valid, it means 'mid' is too large.
                high = mid - 1; // Search in the lower half for a smaller side length.
            }
        }

        return ans; // Return the maximum valid side length found.
    }

    /**
     * Helper function to check if there exists any square of a given side length
     * with a sum of elements less than or equal to the threshold.
     *
     * @param prefixSum The precomputed 2D prefix sum array.
     * @param m         The number of rows in the original matrix.
     * @param n         The number of columns in the original matrix.
     * @param side      The current side length to check.
     * @param threshold The maximum allowed sum for a square.
     * @return true if such a square exists, false otherwise.
     */
    private boolean check(int[][] prefixSum, int m, int n, int side, int threshold) {
        // Iterate through all possible bottom-right corners of a square with the given 'side'.
        // The top-left corner of a square with bottom-right corner (r, c) and side 's' is (r-s+1, c-s+1).
        // Thus, the bottom-right corner can range from (side-1, side-1) to (m-1, n-1).
        for (int r = side - 1; r < m; r++) {
            for (int c = side - 1; c < n; c++) {
                // Calculate the sum of the square with bottom-right corner (r, c) and side 'side'.
                // The top-left corner is (r - side + 1, c - side + 1).
                // Using 1-based indexing for prefixSum:
                // The coordinates in prefixSum will be:
                // Bottom-right of the rectangle for sum: (r+1, c+1)
                // Top-left of the rectangle for sum: (r-side+1, c-side+1)
                //
                // Sum of rectangle (r1, c1) to (r2, c2) inclusive using 0-indexed matrix is:
                // prefixSum[r2+1][c2+1] - prefixSum[r1][c2+1] - prefixSum[r2+1][c1] + prefixSum[r1][c1]
                //
                // Here, r1 = r - side + 1, c1 = c - side + 1, r2 = r, c2 = c.
                // So, the sum is:
                // prefixSum[r+1][c+1] - prefixSum[r-side+1][c+1] - prefixSum[r+1][c-side+1] + prefixSum[r-side+1][c-side+1]

                int sum = prefixSum[r + 1][c + 1] - prefixSum[r - side + 1][c + 1] - prefixSum[r + 1][c - side + 1] + prefixSum[r - side + 1][c - side + 1];

                // If the sum of this square is less than or equal to the threshold, we've found a valid square.
                if (sum <= threshold) {
                    return true;
                }
            }
        }
        // If no square of the given 'side' length satisfies the condition, return false.
        return false;
    }
}
```